src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala

```diff
@@ -157,11 +159,49 @@ class DepositServlet(app: EasyDepositApiApp)
       ).getOrRecoverResponse(respond)
   }
 
-  private def postFileItem(uuid: UUID, path: Path, item: FileItem): Try[(Boolean, FileItem)] = {
-    val fullPath = path.resolve(item.name)
-    managed(item.getInputStream)
-      .apply(app.writeDepositFile(_, user.id, uuid, fullPath))
-      .map((_, item))
+  private def uploadFileItem(uuid: UUID, path: Path, uploadItem: FileItem): Try[(Boolean, FileItem)] = {
+    val extensionIsZip = uploadItem.name.matches(".*.g?z(ip)?")
+    lazy val contentTypeIsZip = uploadItem.contentType.exists(_.matches(
+      "(application|multipart)/(x-)?g?zip(-compress(ed)?)?( .*)?"
+    ))
+    logger.debug(s"is it a ZIP? ${ uploadItem.name } : $extensionIsZip; ${ uploadItem.contentType } : $contentTypeIsZip ")
+    if (extensionIsZip || contentTypeIsZip)
+      managed(uploadItem.getInputStream)
+        .apply(uploadZip(uuid, path, uploadItem, _))
+        .map((_, uploadItem))
+    else {
+      val fullPath = path.resolve(uploadItem.name)
+      managed(uploadItem.getInputStream)
+        .apply(app.writeDepositFile(_, user.id, uuid, fullPath))
+        .map((_, uploadItem))
+    }
+  }
+
+  private def uploadZip(uuid: UUID, path: Path, uploadItem: FileItem, uploadInputStream: InputStream): Try[Boolean] = {
+    val zipInputStream = uploadItem.charset
+      .map(charSet => new ZipInputStream(uploadInputStream, Charset.forName(charSet)))
+      .getOrElse(new ZipInputStream(uploadInputStream))
+
+    def handleZipEntry(zipEntry: ZipEntry) = {
+      if (zipEntry.isDirectory) Success(false)
+      else {
+        logger.info(s"extracting ${ zipEntry.getName } size=${ zipEntry.getSize } crc=${ zipEntry.getCrc } method=${ zipEntry.getMethod }")
+        val fullPath = path.resolve(zipEntry.getName)
+        app.writeDepositFile(zipInputStream, user.id, uuid, fullPath)
+      }
+    }
+
+    while (
```

Comment:
this is kind of a weird construction to see. Usually you have some condition in the while loop and then in the body, you do some things with it. However, in your case, you do everything in de 'condition' part and nothing in the body part.

This makes me wonder whether a Stream generator wouldn't suit you better here. Something like this (_not tested it, maybe you need to modify some things here or there_):

```scala
Try {
    Stream.continually(Option(zipInputStream.getNextEntry))
        .takeWhile(_.nonEmpty)
        .collect { case Some(entry) => entry } // we know now that all elements are Some(...), so we can unpack
        .map(handleZipEntry)
        .foreach {
            case Failure(e) if e.isInstanceOf[ZipException] => throw BadRequestException(s"ZIP file is malformed. ${ uploadItem.name } $e")
            case Failure(e) => throw e
            case Success(_) => // do nothing
        }
}.map(_ => false)
```